url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/ops/SingularOps.java

```diff
@@ -0,0 +1,375 @@
+
+            // only swap if the current index is not the smallest
+            if( bigIndex == i)
+                continue;
+
+            if( bigIndex == -1 ) {
+                // there is at least one uncountable singular value.  just stop here
+                break;
+            }
+
+            double tmp = W.get(i,i);
+            W.set(i,i,bigValue);
+            W.set(bigIndex,bigIndex,tmp);
+
+            if( V != null ) {
+                swapRowOrCol(V, tranV, i, bigIndex);
+            }
+
+            if( U != null ) {
+                swapRowOrCol(U, tranU, i, bigIndex);
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Similar to {@link #descendingOrder(org.ejml.data.DenseMatrix64F, boolean, org.ejml.data.DenseMatrix64F, org.ejml.data.DenseMatrix64F, boolean)}
+     * but takes in an array of singular values instead.
+     * </p>
+     *
+     * @param U Matrix. Modified.
+     * @param tranU is U transposed or not.
+     * @param singularValues Array of singular values. Modified.
+     * @param numSingularValues Number of elements in singularValues array
+     * @param V Matrix. Modified.
+     * @param tranV is V transposed or not.
+     */
+    public static void descendingOrder( DenseMatrix64F U , boolean tranU ,
+                                        double singularValues[] ,
+                                        int numSingularValues ,
+                                        DenseMatrix64F V , boolean tranV )
+    {
+//        checkSvdMatrixSize(U, tranU, W, V, tranV);
+
+        for( int i = 0; i < numSingularValues; i++ ) {
+            double bigValue=-1;
+            int bigIndex=-1;
+
+            // find the smallest singular value in the submatrix
+            for( int j = i; j < numSingularValues; j++ ) {
+                double v = singularValues[j];
+
+                if( v > bigValue ) {
+                    bigValue = v;
+                    bigIndex = j;
+                }
+            }
+
+            // only swap if the current index is not the smallest
+            if( bigIndex == i)
+                continue;
+
+            if( bigIndex == -1 ) {
+                // there is at least one uncountable singular value.  just stop here
+                break;
+            }
+
+            double tmp = singularValues[i];
+            singularValues[i] = bigValue;
+            singularValues[bigIndex] = tmp;
+
+            if( V != null ) {
+                swapRowOrCol(V, tranV, i, bigIndex);
+            }
+
+            if( U != null ) {
+                swapRowOrCol(U, tranU, i, bigIndex);
+            }
+        }
+    }
+
+    /**
+     * Checks to see if all the provided matrices are the expected size for an SVD.  If an error is encountered
+     * then an exception is thrown.  This automatically handles compact and non-compact formats
+     */
+    public static void checkSvdMatrixSize(DenseMatrix64F U, boolean tranU, DenseMatrix64F W, DenseMatrix64F V, boolean tranV ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.